pkg/hbone/doubledialer.go

```diff
@@ -0,0 +1,269 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hbone
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"golang.org/x/net/http2"
+)
+
+// NewDialer creates a Dialer that proxies connections over HBONE to the configured proxy.
+func NewDoubleDialer(outerCfg Config, innerTLSConfig *tls.Config) Dialer {
+	var outerTransport *http2.Transport
+
+	if outerCfg.TLS != nil {
+		outerTransport = &http2.Transport{
+			TLSClientConfig: outerCfg.TLS,
+		}
+	} else {
+		outerTransport = &http2.Transport{
+			// For h2c
+			AllowHTTP: true,
+			DialTLSContext: func(ctx context.Context, network, addr string, tlsCfg *tls.Config) (net.Conn, error) {
+				d := net.Dialer{}
+				if outerCfg.Timeout != nil {
+					d.Timeout = *outerCfg.Timeout
+				}
+				return d.Dial(network, addr)
+			},
+		}
+	}
+
+	return &doubleDialer{
+		outerCfg:       outerCfg,
+		innerTLSConfig: innerTLSConfig,
+		outerTransport: outerTransport,
+	}
+}
+
+type doubleDialer struct {
+	outerCfg       Config
+	innerTLSConfig *tls.Config
+	outerTransport *http2.Transport
+}
+
+// DialContext connects to `address` via the HBONE proxy.
+func (d *doubleDialer) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
+	if network != "tcp" {
+		return net.Dial(network, address)
+	}
+	// TODO: use context
+	c, s := net.Pipe()
+	err := d.proxyTo(s, d.outerCfg, address)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (d doubleDialer) Dial(network, address string) (c net.Conn, err error) {
+	return d.DialContext(context.Background(), network, address)
+}
+
+func (d *doubleDialer) proxyTo(conn io.ReadWriteCloser, req Config, address string) error {
```

Comment:
Hmm, wondering why we need this function duplicated for double HBONE. Given `proxyTo` is a generic thing over `io.ReadWriteCloser`, could we not do

```
func dialDouble() {
  tcpCon := net.Dial(target)
  outerC, outerS := net.Pipe()
  outerHbone(outerS)
  innerHbone(outerC)
```

?

Reply:
Ah slightly misunderstood this. The problem is the `d.transport.RoundTrip` is the thing establishing a connection.

So maybe would be more like

```
func dialDouble() {
  tcpCon := net.Dial(target)
  outerC, outerS := net.Pipe()
  hbone(outerS) // outer
  transport := Transport{Dialer: func() { return outerC }}
  hbone(transport) // inner, connect over the outer
}
```

?

Reply:
I can abstract out the initial round trip, but the copying is still custom because there's an extra layer of piping